### PR TITLE
[Semi-Modular] Ports directional meteors/early warning from oldbase

### PR DESCRIPTION
**Pull request by ghost**
## About The Pull Request
This was-
Easier than I expected actually. Most of the code is just copied from TG's because overriding is better than risking sending meteors twice.

Also, turns out meteor code hasn't been touched in 7 years. So the changes from Cit still work almost exactly the same here.

Notes to Address:
- [ ] I want to add a second announcement when meteors actually are going to start spawning. Need advice on how to do it - should I just add a timer at the end of the announce proc for like 5 seconds before spawns, that calls a second announcement?
- [ ] Timer itself seems off by about 20 seconds extra. I don't know how the math there works, someone help ;-;

## How This Contributes To The Skyrat Roleplay Experience
Simply put: random warningless destruction is kind of crummy for everyone.

"But engineers need something to do!!"
This can and should be solved without smashing the entire station's round and freezing/suffocating them.

A warning combined with a rough direction to focus on gives engineers something new to do: the better they can prepare and brace, the less damage they'll need to properly fix.
Crew as well can both help prep, and clear the targeted areas - meaning they can actually avoid a fully depressurized station and rather get maybe one department depressurized

TL;DR giving people time to brace for impact both means less spontaneous destruction, and preparation can be a reward in and of itself by preventing damage

## Proof of Testing

<details>
<summary>Screenshots/Videos</summary>
  I made a really funny (and rushed) video for this but can't post it when git has a 10MB file limit (its 50 lmao whoops), posted in main-chat:

  [Here](https://discord.com/channels/596783386295795713/640744619302453316/1034426429078782003)
  
Let this screenshot serve as the 'I dont want to look at the video' evidence: 

image
Tested with normal/catastrophic, only came from 1 side. Also tested Stray Meteor, and it's not affected by the timer 😎 still have a chance of one single meteor wiping out a spacewalker!
  
</details>

## Changelog

:cl:
add: Nanotrasen has updated their Early Warning system massively, giving the station a valuable minute or two to prepare before meteors, as well as ensuring the station is only hit from one direction! Be wary though, meteors can still come in at sharp angles and hit fairly deep, or teleport via bluespace!
/:cl:

### CODE REVIEW

**Comment by press-mark** [comment]
basedorion

zased.... so.... fvcking.... zased...

**Comment by Acer202** [comment]
Hopefully we can get working meteor shield satellites too! Currently they don't actually do anything.

File: modular_skyrat/modules/events/code/directional_meteors.dm [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+			wave_type = GLOB.meteorsSPOOKY
+		else
+			WARNING("Wave name of [wave_name] not recognised.")
+			kill()
+
+/datum/round_event/meteor_wave/announce(fake)
+	var/directionstring
+	switch(direction)
+		if(NORTH)
+			directionstring = " towards the fore"
```
Comment (Iamgoofball) [inline modular_skyrat/modules/events/code/directional_meteors.dm]: use north/south/east/west, nautical terms aren't human readable

Comment (ghost) [reply modular_skyrat/modules/events/code/directional_meteors.dm]: what? no.

Comment (Iamgoofball) [reply modular_skyrat/modules/events/code/directional_meteors.dm]: go on the server and ask players to correctly tell me what's in the port of tramstation and see what happens

Comment (ghost) [reply modular_skyrat/modules/events/code/directional_meteors.dm]: no.

**Review by tf-4 — CHANGES_REQUESTED** [review]
I hate how many huge sections of code copy-pasted from /tg/ folders to our own that we have and I don't really want to add more. This can surely be modularised better. I don't think there's really a risk of double meteors anyways.

**Comment by ghost** [comment]
> I hate how many huge sections of code copy-pasted from /tg/ folders to our own that we have and I don't really want to add more. This can surely be modularised better. I don't think there's really a risk of double meteors anyways.

I'll try some less copypaste ones later yea; not as used to coding so I went the easy route until reviews could tell me how they'd prefer it

File: code/modules/meteors/meteors.dm [hunk 2]
```diff
@@ -30,6 +30,8 @@ GLOBAL_LIST_INIT(meteorsD, list(/obj/effect/meteor/medium=15, /obj/effect/meteor
 //Meteor spawning global procs
 ///////////////////////////////
 
+//SKYRAT OVERRIDE START - overwritten in modular_skyrat\modules\events\code\directional_meteors.dm
+/*
 /proc/spawn_meteors(number = 10, list/meteortypes)
 	for(var/i in 1 to number)
 		spawn_meteor(meteortypes)
```
Comment (GoldenAlpharex) [inline code/modules/meteors/meteors.dm]: Just make the one-line edits here, rather than copy-pasting the whole thing to the modular file.

File: code/modules/events/meteor_wave.dm [hunk 3]
```diff
@@ -51,9 +51,13 @@
 /datum/round_event/meteor_wave/announce(fake)
 	priority_announce("Meteors have been detected on collision course with the station.", "Meteor Alert", ANNOUNCER_METEORS)
 
+//SKYRAT OVERRIDE START - overwritten in modular_skyrat\modules\events\code\directional_meteors.dm
+/*
 /datum/round_event/meteor_wave/tick()
 	if(ISMULTIPLE(activeFor, 3))
 		spawn_meteors(5, wave_type) //meteor list types defined in gamemode/meteor/meteors.dm
```
Comment (GoldenAlpharex) [inline code/modules/events/meteor_wave.dm]: Ditto.

File: modular_skyrat/modules/events/code/directional_meteors.dm [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+		if(SOUTH)
+			directionstring = " towards the aft"
+		if(EAST)
+			directionstring = " towards starboard"
+		if(WEST)
+			directionstring = " towards port"
+	priority_announce("Meteors have been detected on collision course with the station[directionstring]. Estimated time until impact: [round((start_when * SSevents.wait) / 10, 0.1)] seconds.", "Meteor Alert", "meteors")
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/events/code/directional_meteors.dm]: I think this would be a little better.
```suggestion
	var/direction_string
	switch(direction)
		if(NORTH)
			direction_string = "fore"
		if(SOUTH)
			direction_string = "aft"
		if(EAST)
			direction_string = "starboard"
		if(WEST)
			direction_string = "port"
	priority_announce("Meteors have been detected on collision course with the station, headed towards its [direction_string]. Estimated time until impact: [round((start_when * SSevents.wait) / 10, 0.1)] seconds.", "Meteor Alert", "meteors")
```

**Review by GoldenAlpharex — CHANGES_REQUESTED** [review]
Copy-pasting is so much more efforts than just making that one-line edit.

File: modular_skyrat/modules/events/code/directional_meteors.dm [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+	switch(direction)
+		if(NORTH)
+			directionstring = "Fore"
+		if(SOUTH)
+			directionstring = "Aft"
+		if(EAST)
+			directionstring = "Starboard"
+		if(WEST)
+			directionstring = "Port"
+	priority_announce("Meteors have been detected on collision course with the station, headed towards its [directionstring]. Estimated time until impact: [round((start_when * SSevents.wait) / 10, 0.1)] seconds.", "Meteor Alert", "meteors")
```
Comment (ghost) [inline modular_skyrat/modules/events/code/directional_meteors.dm]:
```suggestion
	priority_announce("Meteors have been detected on collision course with the station, headed towards its [directionstring] side. Estimated time until impact: [round((start_when * SSevents.wait) / 10, 0.1)] seconds.", "Meteor Alert", "meteors")
```

**Comment by ghost** [comment]
> basedorion
> 
> zased.... so.... fvcking.... zased...

honestly dont wanna finish this PR because it means I dont see this lovely image any more.,..
it keeps me up at night. not out of fear, but out of pride. for i know i am good. i know they care. they love me! they love me!

File: modular_skyrat/modules/events/code/directional_meteors.dm [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+/ ALSO INCLUDES CHANGES IN code/modules/meteors/meteors.dm and code/modules/events/meteor_wave.dm (Mostly just adding direction variables into spots they're needed)
+/ Ctrl+F "//SKYRAT EDIT - DIRECTIONAL METEORS" to find them all
+*/
+
+/datum/round_event/meteor_wave
+	/// Which cardinal direction the meteors will spawn on
+	var/direction
+	/// Cardinal direction translated into Nautical direction, used in announcements and selected on setup()
+	var/directionstring
```
Comment (tf-4) [inline modular_skyrat/modules/events/code/directional_meteors.dm:14]:
```suggestion
	var/direction_string
```

**Comment by ghost** [comment]
TG is messing with dir stuff and I can't be assed so 🤷
